src/storage/src/upsert/types.rs

```diff
@@ -882,6 +911,134 @@ where
         Ok(())
     }
 
+    /// Consolidate the updates into the state during snapshotting. This method requires the
+    /// backend has support for the `multi_merge` operation, and will panic if
+    /// `self.inner.supports_merge()` was not checked before calling this method.
+    /// `multi_merge` will write the updates as 'merge operands' to the backend, and then the
+    /// backend will consolidate those updates with any existing state using the
+    /// `snapshot_merge_function`.
+    ///
+    /// This method can have significant performance benefits over the
+    /// read-then-write method of `consolidate_snapshot_read_write_inner`.
+    async fn consolidate_snapshot_merge_inner<U>(
+        &mut self,
+        updates: U,
+    ) -> Result<SnapshotStats, anyhow::Error>
+    where
+        U: IntoIterator<Item = (UpsertKey, UpsertValue, mz_repr::Diff)> + ExactSizeIterator,
+    {
+        let mut updates = updates.into_iter().peekable();
+
+        let mut stats = SnapshotStats::default();
+
+        if updates.peek().is_some() {
+            let m_stats = self
+                .inner
+                .multi_merge(updates.map(|(k, v, diff)| {
+                    // Transform into a `StateValue<O>` that can be used by the `snapshot_merge_function`
+                    // to merge with any existing value for the key.
+                    let mut val: StateValue<O> = Default::default();
+                    val.merge_update(v, diff, self.bincode_opts, &mut self.bincode_buffer);
+
+                    stats.updates += 1;
+                    if diff > 0 {
+                        stats.inserts += 1;
+                    } else if diff < 0 {
+                        stats.deletes += 1;
+                    }
+
+                    // To keep track of the overall `values_diff` we can use the sum of diffs which
+                    // should be equal to the number of non-tombstoned values in the backend.
+                    // This is a bit misleading as this represents the eventual state after the
+                    // `snapshot_merge_function` has been called to merge all the updates,
+                    // and not the state after this `multi_merge` call.
+                    stats.values_diff += diff;
+
+                    (k, val)
+                }))
+                .await?;
+
+            // This is the total size of the merge values written to the backend, so is an accurate
+            // view of the size stored by the backend after this `multi_merge` operation, but not
+            // representative of the ultimate size of the backend after these merge operands have
+            // been mergedd by the `snapshot_merge_function`.
+            let size: i64 = m_stats.size_written.try_into().expect("less than i64 size");
```

Comment:
The size written is really supposed to be the byte size _of the working set_, not the full dataset size; when you test this, youll see this is much bigger, if used on a workload that has updates (and isnt just a singular snapshot round)

I think the only way we can get a reasonable number is to ignore things that consolidate into tombstones (diff == 0) and instead report the size written as the `sum[u: update](u.diff * u.size)`

Doing this requires a small adjustment to the rocksdb crate, namely that you must pass the diff in so that it can multiply it by the encoded size. You'll have slight inaccuracies as your cancel out the non-value overhead of `StateValue`, but the final size will be reasonable. Let me know if you want to hop on a call and talk about this!

Reply:
makes sense -- just pushed a commit that implements this